types/d3-3d/index.d.ts

```diff
@@ -0,0 +1,338 @@
+import type { SvgProperties } from "csstype";
+
+/**
+ * A generic type for a 3D generator function, capable of transforming data into 3D shapes.
+ */
+export interface Shape3DGenerator<ShapeInput, Shape = ShapeInput> {
+    /**
+     * Generates 3D shapes based on specified parameters and transformations.
+     *
+     * @param data The data representing the 3D shapes.
+     * @returns An array of 3D shapes generated with the specified parameters and transformations.
+     */
+    (data: ShapeInput[]): Shape[];
+
+    /**
+     * Sets or retrieves the origin for rendering the 3D shapes.
+     *
+     * @param o The origin point for rendering the 3D shapes.
+     * @returns If no argument is provided, returns the current origin. Otherwise, sets the origin and returns the function.
+     */
+    origin(o: Coordinate2D): Shape3DGenerator<ShapeInput, Shape>;
+    origin(): Coordinate2D;
```

Comment:
Nit, but I think it's most common to put the no-arg version first.

Reply:
Actually, the order is based on the main usage it is intended for, which is building the shape using these methods.